UnitySDK/Assets/ML-Agents/Scripts/Sensor/WriteAdapter.cs

```diff
@@ -80,26 +88,21 @@ public void SetTarget(TensorProxy tensorProxy, int[] shape, int batchIndex, int
             {
                 if (m_Data != null)
                 {
-                    var height = m_Shape[0];
-                    var width = m_Shape[1];
-                    var channels = m_Shape[2];
-
-                    if (h < 0 || h >= height)
+                    if (h < 0 || h >= m_TensorShape.height)
                     {
-                        throw new IndexOutOfRangeException($"height value {h} must be in range [0, {height-1}]");
+                        throw new IndexOutOfRangeException($"height value {h} must be in range [0, {m_TensorShape.height-1}]");
                     }
-                    if (w < 0 || w >= width)
+                    if (w < 0 || w >= m_TensorShape.width)
                     {
-                        throw new IndexOutOfRangeException($"width value {w} must be in range [0, {width-1}]");
+                        throw new IndexOutOfRangeException($"width value {w} must be in range [0, {m_TensorShape.width-1}]");
                     }
-                    if (ch < 0 || ch >= channels)
+                    if (ch < 0 || ch >= m_TensorShape.channels)
                     {
-                        throw new IndexOutOfRangeException($"channel value {ch} must be in range [0, {channels-1}]");
+                        throw new IndexOutOfRangeException($"channel value {ch} must be in range [0, {m_TensorShape.channels-1}]");
                     }
 
-                    // Math copied from TensorShape.Index(). Note that m_Batch should always be 0
-                    var index = m_Batch * height * width * channels + h * width * channels + w * channels + ch;
-                    m_Data[index + m_Offset] = value;
+                    var index = m_TensorShape.Index(m_Batch, h, w, ch + m_Offset);
+                    m_Data[index] = value;
```

Comment:
Could compute the index for both TensorProxy and float array cases, and use `m_Proxy.data[index] = value;` but I don't see a strong reason for doing that. Same for the vector case.